### DISCUSSION TITLE
How to achieve complex `disableConditions` and `hideConditions`?

### DISCUSSION THREAD

**Discussion by Sidsector9** [woocommerce/woocommerce#44027] [WooCommerce New Product Editor] [answered]
We're handling an extension to add block product editor compatibility.

The goal is to add a checkbox called **Sync** under the **General** group. This checkbox is disabled when the `sku` field is empty. We are able to do this for simple product.

We want to do the same for variation products. The **Sync** checkbox should be enabled if and only if all variation products have `sku` set.

I am trying to use the following as the `disableCondition`

```js
( () => {
    const productData = wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'product', editedProduct.id );

    const filteredProducts = productData.variations.map( variationId => {
        const variationProduct = wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'product', variationId );
        return variationProduct.sku.length === 0;
    });

    return filteredProducts.length > 0;
} )()
```
but I get the following error:

<img width="1326" alt="Screenshot 2024-01-24 at 1 38 23 PM" src="https://github.com/woocommerce/woocommerce/assets/17757960/d1343fba-6743-4e9c-8f2f-6d0215a8f75d">

Is there an alternative way to achieve this at the moment?

**Comment by Sidsector9** [comment]
For more context, this is the function:

```php
/**
 * Adds 'Sync with Square' checkbox field to the product editor under "General" tab.
 *
 * @since x.x.x
 *
 * @param BlockInterface $product_name_block The product name block.
 */
public function add_sync_with_square_field( $product_name_block ) {
	$section = $product_name_block->get_parent();

	$hide_condition = "
		( () => {
			const productData = wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'product', editedProduct.id );
		
			const filteredProducts = productData.variations.map( variationId => {
				const variationProduct = wp.data.select( 'core' ).getEditedEntityRecord( 'postType', 'product', variationId );
				return variationProduct.sku.length === 0;
			});
		
			return filteredProducts.length > 0;
		} )()
	";

	/** @var ProductBlockEditor\ProductTemplates\Section $section */
	$section->add_block(
		array(
			'id'         => self::build_block_id( 'sync-with-square' ),
			'blockName'  => 'woocommerce/product-checkbox-field',
			'attributes' => array(
				'title'    => __( 'Sync with Square', 'woocommerce-square' ),
				'label'    => __( 'Send product data to Square', 'woocommerce-square' ),
				'property' => 'wc_square_synced',
			),
			'disableConditions' => array(
				array(
					'expression' => '"" === editedProduct.sku',
				)
			),
			'hideConditions' => array(
				array(
					'expression' => $hide_condition,
				)
			),
		)
	);
}
```

**Comment by mattsherman** [comment, marked as answer]
@Sidsector9 Hi there!

`hideConditions` and `disableConditions` support "simple" read-only expressions only. Arbitrary JS is not supported. The expressions in `hideConditions` and `disableConditions` are evaluated with a specific grammar that does not support calling functions, creating consts/variables, importing modules, etc.

In order to accomplish what you are aiming to do, you will need to write a custom block for your checkbox. 

We recently published a dev blog post on creating custom blocks for the product editor:

https://developer.woo.com/2024/01/17/getting-to-know-woo-extending-the-new-product-editor-with-react/

**Reply by Sidsector9** [reply]
Makes a lot more sense now. Thanks @mattsherman!
